### PR TITLE
fix(dashboard): dropped terminal viewers no longer pin the PTY as attached forever (#110849, salvage #110859)

### LINKED ISSUE #110849: PtySession.attach() commits attached state before the replay write — a client dropping during attach leaves the PTY permanently unreapable
### Summary

`PtySession.attach()` commits the attached state *before* it writes the replay buffer to the socket, and the caller (`pty_ws`) performs that call **outside** any `try/finally`. A client that drops during the attach replay therefore raises out of the handler with the session already marked `attached=True, last_detached_at=None` and never reaches `PTY_REGISTRY.detach()`.

That state is the one state `reap_idle()` can never reclaim:

* `not s.alive` is `False` — the PTY child is perfectly alive;
* the second clause requires `not s.attached` **and** `s.last_detached_at is not None` — both are false.

`_reap_one_idle_or_raise()` has the same `not attached` requirement, and the 30-minute TTL is anchored to `last_detached_at`, which stays `None` forever. The PTY process group (`node ui-tui/dist/entry.js` + its `python -m tui_gateway.entry` child) then lives indefinitely, holding a tty, the PTY master fd, its loopback sockets and the session's work.

This is an independent entry into "attached forever" — it needs no reverse proxy, no disabled WS ping, and no SIGHUP cascade; the client only has to disappear between `attach` bookkeeping and the first byte of the replay.

### Where (verified on `main` @ a89c1e11)

`hermes_cli/pty_session.py` (attach, ~L98-L114):

```python
if self._ws is not ws:
    await _close_ws(self._ws, WS_CLOSE_SUPERSEDED)
self._ws = ws
self._attach_generation += 1
self.attached = True
self.last_detached_at = None
if snap := self.buffer.snapshot():
    await ws.send_bytes(snap)          # <- raises here, state already committed
```

`hermes_cli/web_routers/chat_ws.py` (L516) — the call site is outside the `try` that owns the detach; `PTY_REGISTRY.detach()` only runs in the writer loop's `finally` further down:

```python
if not await session.attach(ws, force_redraw=not _created):
    await _close_stalled_pty_input(ws, path="keepalive-redraw")
    PTY_REGISTRY.detach(attach_token, ws)
    return
```

### Evidence

Dashboard server log (3 occurrences, same shape), client-side `/api/pty` socket:

```
chat_ws.py:516 in pty_ws
pty_session.py:111 in attach
    await ws.send_bytes(snap)
starlette/websockets.py:89 in send -> raise WebSocketDisconnect(code=1006)
...
uvicorn.protocols.utils.ClientDisconnected
```

Field accumulation on one `hermes dashboard` with a single real terminal (`who` → console + one SSH session):

* peak snapshot: **18** `node ui-tui/dist/entry.js` and **15** `python -m tui_gateway.entry` processes, **15** ttys held, **15** PTY master fds in the dashboard process;
* all ESTABLISHED sockets on the dashboard port were held by the dashboard process itself (no browser clients at all) — i.e. no live viewer anywhere, yet every session stayed attached;
* one orphaned `ui-tui` node (no `tui_gateway` child) survived from the same day five days earlier and was still present in every later snapshot;
* a rescan 45 minutes later showed the total fluctuating (8 nodes / 7 children) but the oldest orphan unchanged — the leak is cumulative, not a one-off.

### Impact

Unbounded process / PTY / fd leak per dropped client, plus indirectly a stuck session lease: a stranded `tui_gateway` with an unfinished turn keeps refreshing its `session_turn_leases` row, which is invisible to watchdog checks keyed on "lease expired" or "owner pid dead" (the expiry is fresh and the pid is alive).

### Proposed fix (any of)

1. In `attach()`, do not commit `_ws`/`attached`/`last_detached_at` until the replay buffer has been delivered — send first, then record the attachment (or roll back in `except`).
2. In `pty_ws`, wrap the `session.attach(...)` call in `try/except` and `PTY_REGISTRY.detach(attach_token, ws)` before re-raising (or fold the attach into the writer loop's `try/finally`).
3. In `reap_idle()`, add a criterion independent of `attached` (e.g. an attach timestamp / socket-liveness probe) so a session whose viewer is gone is eventually reclaimed even if the detach was never observed.

Regression test: a `FakeWS` whose `send_bytes` raises → after the failed attach the session must still be reclaimable (`attached is False` or `last_detached_at is not None`). The existing suite pins the opposite invariant (`tests/test_pty_session.py` — "attached → not reapable") but does not cover an aborted attach.

### Related

* #110164 — orphaned `ui-tui` nodes with dead gateway children; the attach-abort path above is another, proxy-independent source of the same accumulation.
* #96418 — attached session never reaped when the loopback bind disables the WS ping (external trigger); this report is the in-process trigger and does not require the ping to be off.
* #106400 — `reap_idle()` `pop(key)` without default; fixed on `main` (a89c1e11) and **not** the same path as this report.
* PR #110168 — keeps the embedded dashboard TUI alive on SIGHUP; unrelated to the stale-attached state.

### PR DESCRIPTION
**Pull request by teknium1**
A dashboard terminal whose browser drops during the attach replay (or mid-stream) is detached by the session itself, so the idle reaper can reclaim its PTY instead of leaking `node ui-tui` + `tui_gateway` processes, ttys and fds forever.

Fixes #110849

- `PtySession.attach()`: a `send_bytes` failure on the replay rolls the just-committed attach back via `detach()` and returns `False` instead of raising past the caller's only detach path.
- `PtySession._drain()`: a live-forward send failure detaches the socket it was sent to instead of `except Exception: pass`; `detach()`'s existing "only the current socket may detach" guard means a failure against a superseded socket leaves the replacement viewer attached.
- `chat_ws.py::pty_ws`: `attach()` returning `False` now distinguishes "client dropped, already detached" (no socket to close) from "force-redraw write stalled" (close with 1013 as before); the registry detach stays as the guard for both.
- Tests trimmed to two invariants, both red on `origin/main`: a mid-replay drop leaves the session `reap_idle()`-reclaimable; a drain send failure detaches the current socket but never a replacement that attached during the send.

**Live repro:** `.venv/bin/python /tmp/batch111/statedb/probe_pty_attach.py <worktree>` — fake ws whose `send_bytes` raises `starlette.websockets.WebSocketDisconnect(1006)` during attach / drain, then `reap_idle(now=+∞)`:

| | origin/main | this branch |
|---|---|---|
| attach replay drop | `attach raised WebSocketDisconnect; attached=True last_detached_at=None reaped_by_reap_idle=False` | `attach returned False; attached=False reaped_by_reap_idle=True bridge_closed=True` |
| drain send drop | `attached=True last_detached_at=None reaped_by_reap_idle=False` | `attached=False reaped_by_reap_idle=True` |
| superseded socket fails while a live one is attached | `attached=True ws_is_live=True` (unchanged) | `attached=True ws_is_live=True still_registered=True` |

`scripts/run_tests.sh tests/hermes_cli/test_pty_session.py tests/hermes_cli/test_web_server_pty_reconnect.py tests/hermes_cli/test_web_server.py` → 210 passed.

Root cause: `attach()` committed `_ws`/`attached=True`/`last_detached_at=None` before `await ws.send_bytes(snap)`, and the caller's `PTY_REGISTRY.detach()` only runs in the writer loop's `finally` that the raise never reaches, leaving the one state (`alive`, `attached`, no detach timestamp) that neither `reap_idle()` nor `_reap_one_idle_or_raise()` can reclaim.

## Credit

Salvages #110859 by @MohamadKanso (both commits cherry-picked: attach rollback + drain-side detach). #110861 by @liuhao1024 and #110863 by @fangliquanflq reached the same attach-side conclusion independently; #110863 also covered the drain path with an explicit generation guard — here the existing `detach()` identity guard gives the same protection without a second detach helper.

## Infographic

pty-attach-detach-on-send-failure

### Review fixes

- **fixed** (nit) — the stalled-redraw close in `pty_ws` is now gated on socket identity (`session._ws is ws`) instead of `session.attached`. A viewer superseded mid-replay no longer logs "pty input stalled … recycling terminal session" or issues a second close on itself; the replacement viewer is untouched. Reproduced with the reviewer race probe (warning + double close before, neither after); no new test — the probe needs a held `send_bytes` to interleave two attaches, and the existing pty/web_server suites (16 + 4 + 190) pass unchanged.

### CODE REVIEW

**Review by kvnloo — COMMENTED** [review]
Approved — the identity guard in `detach()` is doing exactly the work the body claims, so the superseded-socket case needs no second helper, and the `chat_ws` split between "already detached" and "stalled redraw" reads correctly. The credit section naming the independent convergences is appreciated.

Two non-blocking nits:

1. The replay send is guarded but the `force_redraw` write one line down isn't: `return await self.write(ws, TUI_FORCE_REDRAW)` can still raise out of `attach()` with `attached=True` committed and the caller's writer loop never reached — the same leak class this PR fixes on the replay path. A `try/except → detach → return False` there for symmetry (or a comment on why it can't raise) would close the hole.
2. The superseded-socket no-op case is unit-tested for `_drain` but not for the replay path. Same guard covers both, so it passes today — just noting a regression there would be silent.

Happy to see this land as-is.

**Comment by MohamadKanso** [comment]
small follow up for the redraw edge case is up in #111693

it catches the force redraw write raising before the handler gets to cleanup and keeps the replacement socket safe too

added focused tests for both paths
